### ISSUE TITLE
Feature Request: Make CGAL/OpenCSG render type available to the code. [$100]

### ISSUE THREAD

**Issue by daid** [openscad/openscad#149]
When making my objects in OpenSCAD. I use the OpenCSG render for a "quick render" and CGAL for export, as most people will do.

Now, I want my circles to be pretty detailed for the CGAL export, so I have a  line "$fs=0.4;$fa=5;" at the top. However, this slows down the OpenCSG render a lot. So I comment it out while I am doing quick refreshes with OpenCSG.

Now it would be nice if there was some special variable that was set so you know what type of rendering is used, and allows you to make different code paths depending on that variable.
##

**Comment by nophead** [comment]
>Perhaps an environment array of some sort would make sense, e.g. set $ENV["rendertype"] to "Preview" or "Render" ?

Seems complicated because it looks like an associative array and OpenSCAD doesn't have them. It is also ugly, so I would end up assigning it to a $ variable anyway. Doing so lets it be overridden in a subtree, which might be useful.

 >Does a program want to distinguish between F6 (via menu/cmd-line) or via render()?

Not in my use case no.  I use F5 to draw an assembly view and F6 to lay out the STLs for printing. My assembly view has render()s in all the top level objects, just before I colour them.

>Also variable for GUI v's cmd-line??

Not needed because you can make your own with -D.

**Comment by MichaelAtOz** [comment]
Given the potential for other environmental values, as per #380, what about a function?
You can still do a $ variable,
`$preview=system("preview");`

[Extensible](https://en.wikipedia.org/wiki/Extensibility).

I agree, "preview" should be true within render(), there may be a use for also knowing you are within a render() on top of that, but that's a separate issue.

My use thoughts first went to different $fn etc resolution depending on what state you are in.

**Comment by nophead** [comment]
In my mind $variables are extensible, so I added one more :)

I don't see how one more is a problem. Even if somebody already had used $preview it will simply override the default, so it is fully backwards compatible.

I am not sure why you want to know you are inside render(). You can always put $variables in your code to affect subtrees. This PR solves a different problem. My files are always an assembly view, so it never makes sense to F6 them. To see the printed parts laid out for printing I need to change one line and then press F6. If I accidentally press F6 CGAL goes away for ever trying to make a union of an entire assembly. Now I have

if($preview)
    display_assembly();
else {
    display_box_stl();

    translate([0, box_width + 2, 0])
        display_box_lid_stl();
}

F5 gives this:

![display_mount_f5](https://cloud.githubusercontent.com/assets/566149/23329888/b387f61a-fb39-11e6-8de6-b39a8206f23d.png)

F6 gives;

![display_mount_f6](https://cloud.githubusercontent.com/assets/566149/23329895/e3fb9ad6-fb39-11e6-83a7-69fc05983c40.png)

I have needed this for years as every file I work on I have to change one line between F5 and F6 and remember to put it in preview state if I check it into source control. I put a bounty on it but nobody took the bait.

**Comment by MichaelAtOz** [comment]
> In my mind $variables are extensible, so I added one more :)

MAITRE D: It's only wafer thin.
MR. CREOSOTE: Look. I couldn't eat another thing. I'm absolutely stuffed. Bugger off.
MAITRE D: Oh, sir, just-- just one.
MR. CREOSOTE: [groaning] All right. Just one.
MAITRE D: Just the one, monsieur. Voila.
MR. CREOSOTE: [groaning]
MAITRE D: Bon appetit.
MR. CREOSOTE: [groaning] 

:)

Well, if you need it, next time we can do it right...

**Comment by nophead** [comment]
What is the problem with 100 new $variables? Will OpenSCAD explode, I think not. I don't see why this is not doing it right. Other scripting languages have far more built in variables.

**Comment by MichaelAtOz** [comment]
> Well, if you need it, next time we can do it right...

While sarcastic, I was not being facetious.

I could debate "the problem with 100 new $variables", of which there are many.

So, one more today, hey you need it. Go for it.

It's not like the #380 environment values are going to be implemented any time soon...

**Comment by superjamie** [comment]
Thanks for doing this. Should be a neat way to speed up previews, eg:

~~~cpp
if ($preview == true) {
    $fn = 12;
} else {
    $fn = 360;
};
~~~

**Comment by nophead** [comment]
Yes  but you actually need to do this:
$fn = $preview ? 12 : 360;

The values you assign in an if statement only last until the next closing
curly brace.

**Comment by superjamie** [comment]
Ah, drat. Here's me trying to be explicit and readable and I miss variable scope. I still have a lot of learning to do about OpenSCAD. Thanks again :+1:

**Comment by nophead** [comment]
Solved by #1963
